gosa-core/contrib/gosa.conf.5
.TH gosa.conf 5 "2008-04-07" "GOsa v2.6" "Debian"
.SH NAME
gosa.conf - GOsa configuration file
.SH DESCRIPTION
The gosa.conf file contains configuration information for
.IR GOsa,
a powerful GPL'ed framework for managing accounts and systems in
LDAP databases.
.PP
The gosa.conf file is a XML style configuration file. It is parsed by
the GOsa web application during log in.  The file may contain
extra tabs and newlines for formatting purposes.  Tag keywords in the
file are case-insensitive. Comments should be placed outside of XML
tags and should be encapsulated inside of <!-- --> tags.
.PP
The gosa.conf file can be used to configure the look and feel, behaviour
and access control of the GOsa webinterface.
.SH Configuration layout

The configuration has to be specified inside of the <conf> tags. It
basically consists of three main parts: menu definition, definition
of subdialogs (tabbed dialogs) and the main configuration - including
information about several locations.

.B Layout example:

.nf
  <?xml version="1.0"?>
  
  <conf configVersion="...." >
    <!-- Menu definition -->
    <menu>
    ...
    </menu>
  
    <!-- Tabbed dialog definitions -->
    ...
  
    <!-- Global setup -->
    <main>
  
       <!-- Location specific setups -->
       <location name="">
         ...
       </location>
  
    </main>
  
  </conf>
.fi

.SH Menu definition

This tag defines the side and icon menu inside the
interface. Defining an entry here is no guarantie to get it shown,
though. Only entries with matching ACL's get shown.

There are two types of entries inside of the menu: section and plugin

.B Defining a section

Open a 
.I <section>
tag including a 
.I name
attribute. This will show up in the menu as a new section later on.
Own entries are not handled via I18N by default. Close the 
.I </section>
tag after your plugin definitions.

.B Defining a plugin

Open a 
.I <plugin>
tag including a 
.I "class"
attribute. The 
.I "class" 
should be present inside your GOsa setup - the entry will be ignored if it is not.

Plugins should have an 
.I "acl"
entry, that allows GOsa to decide wether a user is allowed to see a plugin or not.
The 
.I "acl"
string matches with an ACL definition done inside of GOsa.

You can override an icon by specifying the 
.I "icon"
attribute.

For every plugin, you can provide at least seven additional hooks: 
.I precreate,
.I preremove,
.I premodify
.I postcreate,
.I postremove,
.I postmodify
and
.I check.
These can be used to perform special actions when a plugins gets
a create, delete, modify or check request. As a parameter, these
keywords get a shell script or program to the task.

.I The
.B create / delete / modify
.I keywords

These keywords take a full executable path of a script. You can
provide certain parameters in form of LDAP attributes. '%uid'
will pass the current user id, '%dn' the current object dn, etc.

The script gets executed before(pre) and after(post) create, delete or modify tasks.

.I The
.B check
.I keyword

This keyword takes a full executable path of a script. Check is
triggered after you press the
-I "Apply"
or
-I "OK"
button. The complete LDAP entry as it will be written to the
LDAP is passed to your script. If parts of the entry do not
match some logic of your script, just print an error message
to STDOUT. GOsa will show this message and abort the current
process of saving the entry to the LDAP.

.B Example menu definition:

.nf
  <menu>
    <section name="My account">
      <plugin acl="users/user:self" class="user" check="/usr/local/bin/test_user.sh" />
      <plugin acl="users/samba:self" class="sambaAccount" postcreate="/usr/local/bin/create_share '%uid'" />
    </section>
  </menu>
.fi

.SH Tabbed dialog definitions

Tab definitions define the sub plugins which get included for certain
tabbed dialogs. If you change something here, never (!) remove the
primary (the first) "tab" tag which is defined. Most tabbed dialogs
need a primary plugin.

.I "*tab"
should be looked for by a defined plugin. This one will take
every 
.I "tab"
defined
.I "class"
and will show it inside of a tabbed dialog
with the header defined in
.I "name".

.B Example tabbed dialog definition:

.nf
  <grouptabs>
    <tab class="group" name="Generic" />
    <tab class="environment" name="Environment" />
    <tab class="appgroup" name="Applications" />
    <tab class="mailgroup" name="Mail" />
  </grouptabs>
.fi

.SH Main section

The main section defines global settings, which might be overridden by
each location definition inside of this global definition.

.B Example layout:

.nf
  <main default="Example Net"
        listSummary="false"
        ... >

        <location name="Example Net"
                  hash="md5"
                  accountPrimaryAttribute="cn"
                  ...

                  <referral uri="ldaps://ldap.example.net:636/dc=example,dc=net"
                            admin="cn=gosa-admin,dc=example,dc=net"
                            password="secret" />

        </location>

  </main>

.fi

.PP
.B Generic options

.PP
.B forceGlobals
.I bool
.PP
The
.I forceGlobals
statement enables PHP security checks to force register_global settings to
be switched off.
.PP

.B forceSSL
.I bool
.PP
The
.I forceSSL
statement enables PHP security checks to force encrypted access to the web
interface. GOsa will try to redirect to the same URL - just with https://.
.PP

.B warnSSL
.I bool
.PP
The
.I warnSSL
statement enables PHP security checks to detect non encrypted access to
the web interface. GOsa will display a warning in this case.
.PP

.B modificationDetectionAttribute
.I string
.PP
The
.I modificationDetectionAttribute
statement enables GOsa to check if a entry currently being edited has
been modified from someone else outside GOsa in the meantime. It will
display an informative dialog then. It can be set to
.I entryCSN
for OpenLDAP based systems or
.I contextCSN
for Sun DS based systems.
.PP

.B logging
.I string
.PP
The
.I logging
statement enables event logging on GOsa side. Setting it to 
.I true,
GOsa will log every action a user performs via syslog. If you use
rsyslog and configure it to mysql logging, you can browse all events
within GOsa.

GOsa will not log anything, if the logging value is empty or set to
false.
.PP

.B loginAttribute
.I string
.PP
The
.I loginAttribute
statement tells GOsa which LDAP attribute is used as the login name
during login. It can be set to
.I uid, mail
or
.I both.
.PP

.B copyPaste
.I bool
.PP
The
.I copyPaste
statement enables copy and paste for LDAP entries managed with GOsa.
.PP

.B enableSnapshots
.I bool
.PP
The
.I enableSnapshots
statement enables a snapshot mechaism in GOsa. This enables you to save
certain states of entries and restore them later on.
.PP

.B snapshotBase
.I dn
.PP
The
.I snapshotBase
statement defines the base where snapshots should be stored inside of
the LDAP.
.PP

.B snapshotURI
.I uri 
.PP
The
.I snapshotURI
variable defines the LDAP URI for the server which is used to do object
snapshots.
.PP

.B snapshotAdminDn
.I dn
.PP
The
.I snapshotAdminDn
variable defines the user which is used to authenticate when connecting
to
.I snapshotURI.
.PP

.B snapshotAdminPassword
.I string
.PP
The
.I snapshotAdminPassword
variable defines the credentials which are used in combination with
.I snapshotAdminDn
and
.I snapshotURI
in order to authenticate.
.PP

.B config
.I dn
.PP
The
.I config
statement defines the LDAP base, where GOsa stores management information,
such as site wide locking and user notifications.
.PP

.B templateCompileDirectory
.I path
.PP
The
.I templateCompileDirectory
statements defines the path, where the PHP templating engins
.I smarty
should store its compiled GOsa templates for improved speed. This path
needs to be writeable by the user your webserver is running with.
.PP

.B timezone
.I string
.PP
The
.I timezone
statements defines the timezone used inside of GOsa to handle date
related tasks, such as password expiery, vacation messages, etc.
The
.I timezone
value should be a unix conform timezone value like in /etc/timezone.
.PP

.B honourIvbbAttributes
.I bool
.PP
The
.I honourIvbbAttributes
statement enables the IVBB mode inside of GOsa. You need the ivbb.schema
file from used by german authorities.
.PP

.B strictNamingRules
.I bool
.PP
The
.I strictNamingRules
statement enables strict checking of uids and group names. If you need
characters like . or - inside of your accounts, set this to
.I false.
.PP

.B allowUidProposalModification
.I bool
.PP
The
.I allowUidProposalModification
statement enables the abilitiy to modify uid proposals when creating a new user from a template.
.PP

.B honourUnitTags
.I bool
.PP
The
.I honourUnitTags
statement enables checking of
.I unitTag
attributes when using administrative units. If this is set to
.I true
GOsa can only see objects inside the administrative unit a
user is logged into.
.PP

.B rfc2307bis
.I bool
.PP
The
.I rfc2307bis
statement enables rfc2307bis style groups in GOsa. You can use
.I member
attributes instead of memberUid in this case. To make it work
on unix systems, you've to adjust your NSS configuration to
use rfc2307bis style groups, too.
.PP

.B ppdPath
.I path
.PP
The
.I ppdPath
variable defines where to store PPD files for the GOto environment plugins.
.PP

.B ppdGzip
.I bool
.PP
The
.I ppdGzip
variable enables PPD file compression.
.PP

.B resolutions
.I path
.PP
The
.I resolutions
variable defines a plain text file which contains additional resolutions
to be shown in the environment and system plugins.
.PP

.B htaccessAuthentication
.I bool
.PP
The
.I htaccessAuthentication
variable tells GOsa to use either htaccess authentication or LDAP authentication. This
can be used if you want to use i.e. kerberos to authenticate the users.
.PP

.B gosaSupportURI
.I URI
.PP
The
.I gosaSupportURI
defines the major gosa-si server host and the password for GOsa to connect to it.
The password is not required if TLS certificates are provided and the gosa-si
uses TLS.

The format is:

.nf
credentials@host:port
.fi
.PP

.B caCertificate
.I path
.PP
The path of the PEM certificate file of the CA used to sign the
certificates used for securing TLS connections of gosa-si.
Multiple paths separated by spaces may be listed but only the first
path will be used by GOsa² at this time. However,
.B sibridge(1)
will use all of them.
.PP

.B certificate
.I path
.PP
The path of the PEM certificate file to use for authenticating with the gosa-si
server configured via
.I gosaSupportURI
instead of using a password.
.PP

.B keyfile
.I path
.PP
The path of the PEM key file corresponding to the
.I certificate
for authenticating with the gosa-si.
.PP

.B gosaSupportTimeout
.I integer
.PP
The
.I gosaSupportTimeout
sets a connection timeout for all gosa-si actions. See 
.I gosaSupportURI
for details.
.PP

.B Browser and display options

.B listSummary
.I true/false
.PP
The
.I listSummary
statement determines whether a status bar will be shown on the bottom of
GOsa generated lists, displaying a short summary of type and number of
elements in the list.
.PP

.B sendCompressedOutput
.I true/false
.PP
The
.I sendCompressedOutput
statement determines whether PHP should send compressed HTML pages to
browsers or not. This may increase or decrease the performance, depending
on your network.
.PP

.B storeFilterSettings
.I true/false
.PP
The
.I storeFilterSettings
statement determines whether GOsa should store filter and plugin settings
inside of a cookie.
.PP

.B language
.I string
.PP
The
.I language
statement defines the default language used by GOsa. Normally GOsa autodetects
the language from the browser settings. If this is not working or you want to
force the language, just add the language code (i.e. de for german) here.
.PP

.B theme
.I string
.PP
The
.I theme
statement defines what theme is used to display GOsa pages. You can install some
corporate identity like theme and/or modify certain templates to fit your needs
within themes. Take a look at the GOsa
.I FAQ
for more information.
.PP

.B sessionLifetime
.I int
.PP
The
.I sessionLifetime
value defines when a session will expire in seconds. For Debian systems, this will
not work because the sessions will be removed by a cron job instead. Please modify
the value inside of your php.ini instead.
.PP

.B Password options
.PP
.B passwordMinLength
.I integer
.PP
The
.I passwordMinLength
statement determines whether a newly entered password has to be of

debian/changelog
gosa (2.8.225) trusty; urgency=medium

  * sibridge: add commands .description and .gocomment to set the respective attributes

 -- Matthias S. Benkmann <mux2005@gmail.com>  Thu, 28 Jan 2016 11:58:05 +0100

gosa (2.8.224) trusty; urgency=medium

  * sibridge: implemented access controls for TLS connections

 -- Matthias S. Benkmann <mux2005@gmail.com>  Tue, 26 Jan 2016 14:11:19 +0100

gosa (2.8.223) trusty; urgency=medium

  * sibridge: make "decrypt" command actually attempt all keys from the config file as the doc says it does.

 -- Matthias S. Benkmann <mux2005@gmail.com>  Tue, 26 Jan 2016 13:07:41 +0100

gosa (2.8.222) trusty; urgency=medium

  * go-susi: add new TLS certificate permission bit GosaAccessJobs.userMsg.

 -- Matthias S. Benkmann <mux2005@gmail.com>  Tue, 26 Jan 2016 12:58:20 +0100

gosa (2.8.221) trusty; urgency=medium

  * go-susi: parsing of GOsa certificate extensions finished

 -- Matthias S. Benkmann <mux2005@gmail.com>  Tue, 26 Jan 2016 10:28:29 +0100

gosa (2.8.220) trusty; urgency=medium

  * go-susi: parse gosaConnectionLimits certificate extension

 -- Matthias S. Benkmann <mux2005@gmail.com>  Mon, 25 Jan 2016 12:41:19 +0100

gosa (2.8.219) trusty; urgency=medium

  * support multiple caCertificate entries in gosa.conf for use with sibridge; started certificate extension parsing

 -- Matthias S. Benkmann <mux2005@gmail.com>  Mon, 25 Jan 2016 12:02:17 +0100

gosa (2.8.218) trusty; urgency=medium

  * sibridge: fix "panic: Peer() called with my own address. This is a bug!" when using localhost as target

 -- Matthias S. Benkmann <mux2005@gmail.com>  Tue, 19 Jan 2016 11:56:19 +0100

gosa (2.8.217) trusty; urgency=medium

  * fai-helpers: debdb: bugfix: A repository that went away after being cached would permanently prevent the package list from being refreshed.
  * Old behaviour: If even a single Release file could not be read for any repository already in the
  * cache, ONLY the cache data would be used (even if other repositories are still up).
  * New behaviour: The package list will always be refreshed even if some Release files
  * are unavailable. Cached data from a repository that goes away will only be used once.
  * If the same repository is unavailable in 2 subsequent calls of debdb update, its package
  * data is gone from the cache.

 -- Matthias S. Benkmann <mux2005@gmail.com>  Tue, 19 Jan 2016 10:58:05 +0100

gosa (2.8.216) trusty; urgency=medium

  * gosa-fai-helpers: generate_package_list: Support "Verbose" environment variable as described in manual

 -- Matthias S. Benkmann <mux2005@gmail.com>  Tue, 19 Jan 2016 08:35:32 +0100

gosa (2.8.215) trusty; urgency=medium

  * gosa-fai-helpers: upgrade php5-cli from Recommends to Depends.
  * I hope this finally prevents Apache from being installed by default.

 -- Matthias S. Benkmann <mux2005@gmail.com>  Tue, 22 Dec 2015 11:58:59 +0100

gosa (2.8.214) trusty; urgency=medium

  * gosa-fai-helpers: upgrade php5-cli from Suggests to Recommends

 -- Matthias S. Benkmann <mux2005@gmail.com>  Tue, 22 Dec 2015 11:47:44 +0100

gosa (2.8.213) trusty; urgency=medium

  * GOsa: implement TLS support when connecting to gosa-si

 -- Matthias S. Benkmann <mux2005@gmail.com>  Mon, 21 Dec 2015 13:56:18 +0100

gosa (2.8.212) trusty; urgency=medium

  * sibridge: -l now takes a port argument, requires TLS and accepts non-localhost connections

 -- Matthias S. Benkmann <mux2005@gmail.com>  Fri, 18 Dec 2015 22:19:36 +0100

gosa (2.8.211) trusty; urgency=medium

  * sibridge: Read config from gosa.conf instead of gosa-si/server.conf
  * This change is necessary because the certificate sibridge needs to use
  * to access go-susi is the one from GOsa, not the one used by the go-susi server
  * itself because the server certificate will not usually have the permission bits
  * set for doing the stuff sibridge needs to do.

 -- Matthias S. Benkmann <mux2005@gmail.com>  Fri, 18 Dec 2015 20:03:21 +0100

gosa (2.8.210) trusty; urgency=medium

  * sibridge: support TLS

 -- Matthias S. Benkmann <mux2005@gmail.com>  Fri, 18 Dec 2015 17:30:19 +0100

gosa (2.8.209) trusty; urgency=medium

  * go-susi: Finally get rid of the errors in the log for various CLMSG_* messages

 -- Matthias S. Benkmann <mux2005@gmail.com>  Fri, 18 Dec 2015 13:14:29 +0100

gosa (2.8.208) trusty; urgency=medium

  * go-susi: add timeout for sending STARTTLS if TLS-only mode; send error reply to peer if STARTTLS is required

 -- Matthias S. Benkmann <mux2005@gmail.com>  Fri, 18 Dec 2015 10:36:53 +0100

gosa (2.8.207) trusty; urgency=medium

  * go-susi: fix brokenness after TLS server support

 -- Matthias S. Benkmann <mux2005@gmail.com>  Fri, 18 Dec 2015 09:54:37 +0100

gosa (2.8.206) trusty; urgency=medium

  * go-susi: TLS support for server-server communication (broken, ATM)

 -- Matthias S. Benkmann <mux2005@gmail.com>  Fri, 11 Dec 2015 19:28:23 +0100

gosa (2.8.205) trusty; urgency=medium

  * go-susi: proper TLS validation of loopback IP even for go-susi --stats

 -- Matthias S. Benkmann <mux2005@gmail.com>  Fri, 11 Dec 2015 16:25:39 +0100

gosa (2.8.204) trusty; urgency=medium

  * go-susi: log to logfile even for go-susi --stats

 -- Matthias S. Benkmann <mux2005@gmail.com>  Fri, 11 Dec 2015 16:01:14 +0100

gosa (2.8.203) trusty; urgency=medium

  * go-susi: translate all loopback addresses to our external address when doing TLS verification
  * This allows a localhost connection to succeed with a certificate for the external IP
  * address and allows a connection from the external IP address to succeed with a
  * certificate for 127.0.0.1

 -- Matthias S. Benkmann <mux2005@gmail.com>  Fri, 11 Dec 2015 14:17:45 +0100

gosa (2.8.202) trusty; urgency=medium

  * go-susi: TLS support for clients should now be complete

 -- Matthias S. Benkmann <mux2005@gmail.com>  Fri, 11 Dec 2015 12:09:54 +0100

gosa (2.8.201) trusty; urgency=medium

  * go-susi: yet another test suite timing fix


gosa-core/html/include/gosa.js
/*
 * This code is part of GOsa (http://www.gosa-project.org)
 * Copyright (C) 2003-2010 GONICUS GmbH
 * 
 * ID: $$Id: index.php 15301 2010-01-26 09:40:08Z cajus $$
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/* Install event handlers */
Event.observe(window, 'resize', resizeHandler);
Event.observe(window, 'load', resizeHandler);
Event.observe(window, 'load', initProgressPie);
Event.observe(window, 'keypress', keyHandler);


/* Ask before switching a plugin with this function */
function question(text, url)
{
	if(document.mainform.ignore || $('pluginModified') == null || $('pluginModified').value == 0){
		location.href= url;
		return true;
	}
	if(confirm(text)){
		location.href= url;
		return true;
	}
	return false;
}


/* Toggle checkbox that matches regex */
function chk_set_all(regex,value)
{
	for (var i = 0; i < document.mainform.elements.length; i++) {
		var _id=document.mainform.elements[i].id;
		if(_id.match(regex)) {
			document.getElementById(_id).checked= value;
		}
	}
}


function toggle_all_(regex,state_object)
{
	state = document.getElementById(state_object).checked;
	chk_set_all(regex, state);
}


/* Scroll down the body frame */
function scrollDown2()
{
	document.body.scrollTop = document.body.scrollHeight - document.body.clientHeight;
}


/* Toggle checkbox that matches regex */
function acl_set_all(regex,value)
{
	for (var i = 0; i < document.mainform.elements.length; i++) {
		var _id=document.mainform.elements[i].id;
		if(_id.match(regex)) {
			document.getElementById(_id).checked= value;
		}
	}
}

/* Toggle checkbox that matches regex */
function acl_toggle_all(regex)
{
    for (var i = 0; i < document.mainform.elements.length; i++) {
        var _id=document.mainform.elements[i].id;
        if(_id.match(regex)) {
            if (document.getElementById(_id).checked == true){
                document.getElementById(_id).checked= false;
            } else {
                document.getElementById(_id).checked= true;
            }
        }
    }
}


/* Global key handler to estimate which element gets the next focus if enter is pressed */
function keyHandler(DnEvents) {

    var element = Event.element(DnEvents);

    // determines whether Netscape or Internet Explorer
    k = (Prototype.Browser.Gecko) ? DnEvents.keyCode : window.event.keyCode;
    if (k == 13 && element.type!='textarea') { // enter key pressed

        // Stop 'Enter' key-press from beeing processed internally
        Event.stop(DnEvents);

        // No nextfield explicitly specified 
        var next_element = null;
        if(typeof(nextfield)!='undefined') {
            next_element = $(nextfield);
        }
       
        // nextfield not given or invalid
        if(!next_element || typeof(nextfield)=='undefined'){
            next_element = getNextInputElement(element);
        }

        if(element != null && element.type == 'submit'){

            // If the current element is of type submit, then submit the button else set focus
            element.click();
            return(false);
        }else if(next_element!=null && next_element.type == 'submit'){
        
            // If next element is of type submit, then submit the button else set focus
            next_element.click();
            return(false);
        }else if(next_element){
            next_element.focus();
            return;
        }
    
    } else if (k==9 && element.type=='textarea') {
        Event.stop(DnEvents);
        element.value += "\t";
        return false;
    }
}

function getNextInputElement(element)
{
    var found = false;
    for (var e=0;e< document.forms.length; e++){
        for (var i = 0; i < document.forms[e].elements.length; i++) {           
            var el = document.forms[e].elements[i]
            if(found && !el.disabled && el.type!='hidden' && !el.name.match(/^submit_tree_base/) && !el.name.match(/^bs_rebase/)){
                return(el);    
            }                                                           
            if((el.id != "" && el.id==element.id) || (el.name != "" && el.name==element.name)){        
                found =true;
            }
        }                                                              
    }                                                              
}

function changeState() {
	for (var i = 0; i < arguments.length; i++) { 
		var element = $(arguments[i]); 
		if (element.hasAttribute('disabled')) { 
			element.removeAttribute('disabled'); 
		} else { 
			element.setAttribute('disabled', 'disabled'); 
		} 
	} 
}

function changeSelectState(triggerField, myField) {
	if (document.getElementById(triggerField).value != 2){
		document.getElementById(myField).disabled= true;
	} else {
		document.getElementById(myField).disabled= false;
	}
}

function changeSubselectState(triggerField, myField) {
	if (document.getElementById(triggerField).checked == true){
		document.getElementById(myField).disabled= false;
	} else {
		document.getElementById(myField).disabled= true;
	}
}

function changeTripleSelectState(firstTriggerField, secondTriggerField, myField) {
	if (
			document.getElementById(firstTriggerField).checked == true &&
			document.getElementById(secondTriggerField).checked == true){
		document.getElementById(myField).disabled= false;
	} else {
		document.getElementById(myField).disabled= true;
	}
}

<!-- Second field must be non-checked -->
function changeTripleSelectState_2nd_neg(firstTriggerField, secondTriggerField, myField) {
	if (
			document.getElementById(firstTriggerField).checked == true &&
			document.getElementById(secondTriggerField).checked == false){
		document.getElementById(myField).disabled= false;
	} else {
		document.getElementById(myField).disabled= true;
	}
}


function popup(target, name) {
	var mypopup= 
		window.open(
				target,
				name,
				"width=600,height=700,location=no,toolbar=no,directories=no,menubar=no,status=no,scrollbars=yes"
			   );
	mypopup.focus();
	return false;
}

function js_check(form) {
	form.javascript.value = 'true';
}

function divGOsa_toggle(element) {
	var cell;
	var cellname="tr_"+(element);

	if (Prototype.Browser.Gecko) {
		document.poppedLayer = document.getElementById(element);
		cell= document.getElementById(cellname);

		if (document.poppedLayer.style.visibility == "visible") {
                        $(element).hide();
			cell.style.height="0px";
			document.poppedLayer.style.height="0px";
		} else {
                        $(element).show();
			document.poppedLayer.style.height="";
			if(document.defaultView) {
				cell.style.height=document.defaultView.getComputedStyle(document.poppedLayer,"").getPropertyValue('height');
			}
		}
	} else if (Prototype.Browser.IE) {
		document.poppedLayer = document.getElementById(element);
		cell= document.getElementById(cellname);
		if (document.poppedLayer.style.visibility == "visible") {
			$(element).hide();
			cell.style.height="0px";
			document.poppedLayer.style.height="0px";
			document.poppedLayer.style.position="absolute";
		} else {
                        $(element).show();
			cell.style.height="";
			document.poppedLayer.style.height="";
			document.poppedLayer.style.position="relative";
		}
	}
}

function resizeHandler (e) {
	if (!e) e= window.event;

    	// This works with FF / IE9. If Apples resolves a bug in webkit,
    	// it works with Safari/Chrome, too.
	if ($("d_scrollbody") && $("t_nscrollbody")) {
      		var contentHeight= document.viewport.getHeight() - 216;
      		if ($$('div.plugin-actions').length != 0) {
        		var height= 0;
        		$$('div.plugin-actions').each(function(s) {
          			height+= s.getHeight();
        		});
        		contentHeight-= height + 25;
      		}

		if (Prototype.Browser.IE) {
	      		document.getElementById('d_scrollbody').style.minHeight = contentHeight+24+'px';
	      		document.getElementById('t_nscrollbody').style.minHeight = contentHeight+'px';
		} else {
	      		document.getElementById('d_scrollbody').style.height = contentHeight+24+'px';
	      		document.getElementById('t_nscrollbody').style.height = contentHeight+'px';
		}
    	}

	return true;
}


function absTop(e) {
	return (e.offsetParent)?e.offsetTop+absTop(e.offsetParent) : e.offsetTop;
}

/* Set focus to first valid input field
   avoid IExplorer warning about hidding or disabled fields
 */
function focus_field()
{
	var i     = 0;
	var e     = 0;
	var found = false;
	var element_name = "";
	var element =null;

	while(focus_field.arguments[i] && !found){

		var tmp = document.getElementsByName(focus_field.arguments[i]);
		for(e = 0 ; e < tmp.length ; e ++ ){

			if(isVisible(tmp[e])){
				found = true;
				element = tmp[e];
				break;
			}
		}
		i++;
	}

	if(element && found){
		element.blur();
		element.focus();
	}
}


/*  This function pops up messages from message queue 
    All messages are hidden in html output (style='display:none;').
    This function makes single messages visible till there are no more dialogs queued.

    hidden inputs: 
    current_msg_dialogs		- Currently visible dialog
    closed_msg_dialogs		- IDs of already closed dialogs 
    pending_msg_dialogs		- Queued dialog IDs. 
 */
function next_msg_dialog()
{
	var s_pending = "";
	var a_pending = new Array();
	var i_id			= 0;
	var i					= 0;
	var tmp				= "";
	var ele 			= null;
	var ele2 			= null;
	var cur_id 		= "";

	if(document.getElementById('current_msg_dialogs')){
		cur_id = document.getElementById('current_msg_dialogs').value;
		if(cur_id != ""){
			ele = document.getElementById('e_layer' + cur_id);
			ele.onmousemove = "";
			$('e_layer' + cur_id).hide();
			document.getElementById('closed_msg_dialogs').value += "," + cur_id;
			document.getElementById('current_msg_dialogs').value= ""; 
		}
	}

	if(document.getElementById('pending_msg_dialogs')){
		s_pending = document.getElementById('pending_msg_dialogs').value;
		a_pending = s_pending.split(",");
		if(a_pending.length){
			i_id = a_pending.pop();
			for (i = 0 ; i < a_pending.length; ++i){
				tmp = tmp + a_pending[i] + ',';
			}
			tmp = tmp.replace(/,$/g,"");
			if(i_id != ""){
				ele = document.getElementById('e_layer' + i_id);
				ele3 = document.getElementById('e_layerTitle' + i_id);
				ele.style.display= 'block'	;
				document.getElementById('pending_msg_dialogs').value= tmp;
				document.getElementById('current_msg_dialogs').value= i_id;
				ele2 = document.getElementById('e_layer2') ;
				ele3.onmousedown = start_move_div_by_cursor;
				ele2.onmouseup 	= stop_move_div_by_cursor;
				ele2.onmousemove = move_div_by_cursor;
			}else{
				ele2 = document.getElementById('e_layer2') ;
				ele2.style.display ="none";
			}
		}
	}
}


/* Drag & drop for message dialogs */
var enable_move_div_by_cursor = false;		// Indicates wheter the div movement is enabled or not 
var mouse_x_on_div	= 0;									// 
var mouse_y_on_div 	= 0;
var div_offset_x  	= 0;
var div_offset_y  	= 0;

/* Activates msg_dialog drag & drop
 * This function is called when clicking on a displayed msg_dialog 
 */
function start_move_div_by_cursor(e)
{
	var x = 0; 
	var y = 0;	
	var cur_id = 0;
	var dialog = null;
	var event = null;

	/* Get current msg_dialog position
	 */
	cur_id = document.getElementById('current_msg_dialogs').value;
	if(cur_id != ""){
		dialog = document.getElementById('e_layer' + cur_id);
		x = dialog.style.left;
		y = dialog.style.top;
		x = x.replace(/[^0-9]/g,"");
		y = y.replace(/[^0-9]/g,"");
		if(!y) y = 1;
		if(!x) x = 1;
	}

	/* Get mouse position within msg_dialog 
	 */
	if(window.event){
		event = window.event;
		if(event.offsetX){
			div_offset_x   = event.clientX -x;
			div_offset_y   = event.clientY -y;
			enable_move_div_by_cursor = true;
		}
	}else if(e){
		event = e;
		if(event.layerX){
			div_offset_x	= event.screenX -x;
			div_offset_y	= event.screenY -y;
			enable_move_div_by_cursor = true;
		}
	}
}


/* Deactivate msg_dialog movement 
 */
function stop_move_div_by_cursor()
{
	mouse_x_on_div = 0;
	mouse_y_on_div = 0;
	div_offset_x = 0;
	div_offset_y = 0;
	enable_move_div_by_cursor = false;
}


/* Move msg_dialog with cursor */
function move_div_by_cursor(e)
{
	var event 				= false;
	var mouse_pos_x		= 0;
	var mouse_pos_y 	= 0;
	var	cur_div_x = 0;
	var cur_div_y = 0;
	var cur_id	= 0;
	var dialog = null;


	if(undefined !== enable_move_div_by_cursor && enable_move_div_by_cursor == true){

		if(document.getElementById('current_msg_dialogs')){

			/* Get mouse position on screen 
			 */
			if(window.event){
				event = window.event;
				mouse_pos_x  =event.clientX;
				mouse_pos_y  =event.clientY;
			}else if (e){
				event = e;
				mouse_pos_x  =event.screenX;
				mouse_pos_y  =event.screenY;
			}else{
				return;
			}

			/* Get id of current msg_dialog */
			cur_id = document.getElementById('current_msg_dialogs').value;
			if(cur_id != ""){
				dialog = document.getElementById('e_layer' + cur_id);

				/* Calculate new position */
				cur_div_x = mouse_pos_x - div_offset_x;
				cur_div_y = mouse_pos_y - div_offset_y;

				/* Ensure that dialog can't be moved out of screen */
				if(cur_div_x < 0 ) cur_div_x = 0
					if(cur_div_y < 0 ) cur_div_y = 0

						/* Assign new values */
						dialog.style.left = (cur_div_x ) + "px";
				dialog.style.top  = (cur_div_y ) + "px";
			}
		}
	}
}


function isVisible(obj)
{
    if (obj == document) return true

    if (!obj) return false
    if (!obj.parentNode) return false
    if (obj.style) {
        if (obj.style.display == 'none') return false
        if (obj.style.visibility == 'hidden') return false
    }

    //Try the computed style in a standard way
    if (window.getComputedStyle) {
        var style = window.getComputedStyle(obj, "")
        if (style.display == 'none') return false
        if (style.visibility == 'hidden') return false
    }

    //Or get the computed style using IE's silly proprietary way
    var style = obj.currentStyle
    if (style) {
        if (style['display'] == 'none') return false
        if (style['visibility'] == 'hidden') return false
    }

    return isVisible(obj.parentNode)
}


/* Check if capslock is enabled */
function capslock(e) {
    e = (e) ? e : window.event;

    var charCode = false;
    if (e.which) {
        charCode = e.which;
    } else if (e.keyCode) {
        charCode = e.keyCode;
    }

    var shifton = false;
    if (e.shiftKey) {
        shifton = e.shiftKey;
    } else if (e.modifiers) {
        shifton = !!(e.modifiers & 4);
    }

    if (charCode >= 97 && charCode <= 122 && shifton) {
        return true;
    }

    if (charCode >= 65 && charCode <= 90 && !shifton) {
        return true;
    }

    return false;
}

function setProgressPie(context, percent)
{
    context.clearRect(0, 0, 22, 22);

    var r = "FF";
    var g = "FF";
    var b = "FF";

    // Fade yellow
    if (percent > 50) {
        d = 255 - parseInt((percent-50) * 255 / 50)
            b = d.toString(16);
    }

    // Fade red
    if (percent > 75) {
        d = 255 - parseInt((percent-75) * 255 / 25)
            g = d.toString(16);
    }

    context.strokeStyle = "#" + r  + g + b
        context.fillStyle = context.strokeStyle;

    context.beginPath();
    context.moveTo(11,11)
        context.arc(11,11,8,-Math.PI/2,-Math.PI/2 + Math.PI*percent/50,true);
    context.closePath();
    context.fill();

    context.moveTo(11,11)
        context.beginPath();
    context.arc(11,11,8,0,Math.PI*2,false);
    context.closePath();
    context.stroke();
}

function initProgressPie(){
    var canvas = $('sTimeout');

    // Check the element is in the DOM and the browser supports canvas
    if(canvas && canvas.getContext) {
        var percent = 0.01;
        var context = canvas.getContext('2d');
        setProgressPie(context, percent);

        // Extract timeout and title string out out canvas.title
        var data = canvas.title;
        var timeout = data.replace(/\|.*$/,'');
        var title = data.replace(/^.*\|/,'');
        var interval = 1;
        var time = 0;
        setInterval(function() {

                // Calculate percentage 
                percent+= (interval / timeout) * 100;

                // Increase current time by interval
                time += interval;

                // Generate title
                var minutes = parseInt((timeout-time) / 60 );
                var seconds = '' + parseInt((timeout-time) % 60);
                if(seconds.length == 1) seconds = '0' + seconds ;
                minutes = minutes + ':' + seconds;

                // Set new  canval title
                canvas.title=  title.replace(/%d/ ,minutes);
                setProgressPie(context, percent);

                if (percent>99) percent= 99;
                }, (interval * 1000));
    }
}

/* Scroll down the body frame */
function scrollDown2()
{
    document.body.scrollTop = document.body.scrollHeight - document.body.clientHeight;
}


// Global storage for baseSelector timer
var rtimer;

// vim:ts=2:syntax
